Plugin events

mark-prins opened the pull request:
<!-- IMPORTANT!
  - Every PR must reference an issue; this helps to explain the intent of the PR
 -->

Fixes #2186 

# 👩🏻‍💻 What does this PR do? 
 <!-- Explain the changes you made, and why they're needed. Add a screenshot if you've made any UI changes!  -->
Adds some methods to the `PluginProvider`

- addEventListener
- dispatchEvent
- removeEventListener

The methods require the `EventType` union type to be specified and the type of plugin, which together determine where in the app an event should be raised and handled.


# 🧪 How has/should this change been tested? 
<!-- Explain how to setup for testing here if it is not already obvious, and how you've tested this PR. -->

There's an example in the plugin example branch of events - where the events are used in the StockDonor plugin to trigger validation on the stock line edit form, and to save data when the line is saved.

## 💌 Any notes for the reviewer?
<!-- eg. Do you have any specific questions for the reviewer? Is there a high risk/complicated change they should focus on? If there are any general areas of the codebase your changes might have have touched or could cause side effects to, mention them here.-->

## 📃 Documentation
<!-- Note down any areas which require documentation updates -->
_No user facing changes_

fergie-nz commented:
I'm curious - I can't see these files in the omSupply repo. 

Is this a separate semi-detached code base?

mark-prins commented on client/packages/common/src/plugins/hooks/usePluginColumns.ts:
```diff
@@ -3,21 +3,22 @@ import { ColumnPluginType } from '../types';
   type: ColumnPluginType;
 }) {
-  const { columnPlugins } = usePluginProvider();
+  const columnPlugins = usePluginProvider(state => state.columnPlugins);
```
using `usePluginProvider()` causes the provider and all children to re-render, so a specific fetch is required.

mark-prins commented on client/packages/common/src/plugins/hooks/usePluginColumns.ts:
```diff
@@ -3,21 +3,22 @@ import { ColumnPluginType } from '../types';
       plugin.column().then(column => {
         const mapped = column as unknown as ColumnDefinition<T>;
-        setPluginColumns([...pluginColumns, mapped]);
+        setPluginColumns(ArrayUtils.uniqBy([...pluginColumns, mapped], 'key'));
```
if the same column is added again, we only want one of them! I've seen this with HMR, but it's possible it will happen under normal conditions

mark-prins commented on client/packages/common/src/plugins/hooks/usePluginEvents.ts:
```diff
@@ -0,0 +1,19 @@
+import { useMemo } from 'react';
+import { usePluginProvider } from '../components';
+
+export const usePluginEvents = () => {
```
added this as I was having a lot of re-render issues; and found the need to use a selector function to get values out of zustand state clumsy. Am worried that people won't use it correctly and will have render issues

mark-prins replied:
The advantage is usually talked about in terms of ease of use. personally, I've found the need to selectively extract rather annoying to use.

mark-prins commented on client/packages/common/src/ui/layout/tables/DataTable.tsx:
```diff
@@ -94,6 +94,12 @@ const DataTableComponent = <T extends RecordWithId>({
 
+  // if the columns array changes, such as when a plugin column is added
+  // reset the display columns to the full set
+  useEffect(() => {
```
the column wasn't showing - likely after I'd fixed a needless re-render

mark-prins commented on client/packages/common/src/ui/layout/tables/components/index.tsx:
```diff
@@ -30,14 +27,26 @@ export const BasicCell = <T extends RecordWithId>({
   </Box>
 );
 
+export const BasicCell = <T extends RecordWithId>({
```
added this, so that I could make use of the basic cell layout without recreating it

mark-prins replied:
In the `StockDonorColumn` I'm using the same cell layout and rather than just duplicating that, I've used the `BasicCellLayout` which is exported here, so:

```
  return <BasicCellLayout>{data?.data ?? ''}</BasicCellLayout>;
```

I wanted the same border radius and padding and text overflow that the basic cell has, so that the new cell looks the same. I also want the plugin cell to be updated if the base properties ever get changed

mark-prins commented on client/packages/system/src/Stock/Components/StockLineEditModal.tsx:
```diff
@@ -60,10 +62,19 @@ export const StockLineEditModal: FC<StockLineEditModalProps> = ({
+  const { dispatchEvent, addEventListener, removeEventListener } =
+    usePluginEvents();
+  const [hasChanged, setHasChanged] = useState(false);
+  const plugins = usePluginElements({
```
pulled the plugins up here, because the form is mutating the `draft` object on every keypress. the plugin, I decided, doesn't need to have the current values, and is ok with the original value of the stockline object

mark-prins commented on client/packages/system/src/Stock/Components/StockLineEditModal.tsx:
```diff
@@ -72,6 +83,14 @@ export const StockLineEditModal: FC<StockLineEditModalProps> = ({
+  const onChange = () => setHasChanged(true);
+
+  useEffect(() => {
+    addEventListener('onChange', 'StockEditForm', onChange);
```
adding this so that the validation is updated when the plugin changes it's own data

mark-prins replied:
the `onCancel` is vaguely what I had in mind for that, it's slightly different than unchanging tho. I couldn't see a need to 'unChange' yet, and I've only implemented the minimum of what I think is useful

mark-prins commented on client/packages/system/src/Stock/Components/StockLineForm.tsx:
```diff
@@ -74,6 +75,12 @@ export const StockLineForm: FC<StockLineFormProps> = ({ draft, onUpdate }) => {
     }
   };
 
+  const debouncedUpdate = useDebounceCallback(
```
added some debounce, as the form is jumpy for me - the mutating of `draft` is causing a render which is noticeable when a plugin is rendered too

mark-prins commented:
> Is this a separate semi-detached code base?

It's in a feature branch for now - waiting until this one is merged before it's ready to bring into develop

clemens-msupply reviewed (changes requested):
There is still some re-render bug. Pressing "d" quickly sends you to the Dashboard:

https://github.com/openmsupply/open-msupply/assets/88299195/ac406765-aca9-4437-8d79-7809bf29a3ac

I actually see some similar bugs for the patient search on the contact-tracing page so it might be some underlying issue.
I set to "Request changes" but it might be another issue. Otherwise it all looks good!

One thing I was wondering about is if we should make it more explicit which events are valid for which `ComponentPluginType`. For example, we could do some TS magic to restrict the event type based on the `ComponentPluginType` or we could define events for every `ComponentPluginType` like "StockEditFormSaved", "StockEditFormChanged". 

Another thing I was thinking about is a seperation between events that are coming from plugins and for events that are comming from the host. We could then only make events that are applicable to plugins visible to the plugin, i.e. which events they can listen to and which event they can subscribe to. This might help when implementing plugins, e.g. don't accidentally send the wrong event which then affects other plugins... One way it could be done is like https://developer.chrome.com/docs/extensions/reference/events/.

Regarding security, we more or less fully trust plugins, don't we? so we don't have to restrict what events a plugin can send, right? For example, a plugin could send an event for a different plugin type.

mark-prins commented on client/packages/common/src/plugins/types.ts:
```diff
@@ -37,6 +37,8 @@ export type ComponentPluginType =
 
 export type ColumnPluginType = 'Stock';
 
+export type EventType = 'onSave' | 'onCancel' | 'onChange';
```
I thought of events as more flexible than that - either the host or the plugin is able to either subscribe to and raise events.

clemens-msupply replied:
Ok might be able to use TypeScript to restrict the EventType base on the plugin area...

Some docs comment about the event types would be useful. For example, `onChange` is confusing to me this is like a command isn't it, like `markAsDirty` isn't it?

clemens-msupply reviewed (approved):
Realized, my re-render complain is actually about the example plugin and not really part of this PR is it? so will approve.

mark-prins commented on client/packages/common/src/plugins/components/PluginProvider.tsx:
```diff
@@ -3,49 +3,106 @@ import {
+  removeEventListener: (
+    eventType: EventType,
+    pluginType: ComponentPluginType,
+    listener: EventListener
```
I've followed the same pattern that javascript uses for events, which is that the listener is only removed if it is an exact match for the object which was added. I thought that this was less limiting than only allowing a single listener for a given combination of eventType and pluginType.. though in practice I think that a single listener is a valid approach.

I'll change to use PluginEventListener and use the types as the match when adding and removing. I think that'll be clearer, thanks

mark-prins commented:
I fixed the re-render issue; it is a problem with the modal not the plugin. The issue is with the `onUpdate` which was causing a lot of renders - when the draft object isn't changed, but the method is called, then a new object is created and the setState causes a render. It's a little clumsy, but to resolve I've added an `isEqual` check.

The problem surfaces because the cost price input has `autoFocus` set, when you click into another input, this fires the `onChange` event for the cost price, calling `onUpdate` and making the form render about 20 times. This is resolved now.

mark-prins commented:
thanks for the review @clemens-msupply!

I've changed the event types to be very specific types (onSaveStockEditForm and onChangeStockEditForm) even though this is not my preference. I would have the more standard `onChange` and `onSave` and have the potential for dispatching events for which a handler has not been implemented. 

The change, I think, addresses the concern about limiting a type of event being dispatched for an incorrect type of plugin - though it does still allow an event to be dispatched for a different plugin (the StockDonorEditInput component could fire an event for 'onChangeOutboundShipmentEditForm' if such a thing existed)

clemens-msupply commented:
thanks for the changes!

> I've changed the event types to be very specific types (onSaveStockEditForm and onChangeStockEditForm) even though this is not my preference. I would have the more standard `onChange` and `onSave` and have the potential for dispatching events for which a handler has not been implemented.

I am open for it. We just need to document which events can happen for which plugins, etc... Think we can even enforce this using TS, e.g. `addListener<'Dashboard'>(...)`
